### DISCUSSION TITLE
Wakeup by USART (SerialDeepSleep example from STM32LowPower)

### DISCUSSION THREAD

**Discussion by tshcherban** [stm32duino/Arduino_Core_STM32#2398] [General]
Hi
I've looked at the `SerialDeepSleep` example from the `STM32LowPower` library and found out that USART is losing an incoming data. MCU is woken up from `sleep`/`deepSleep` modes but `Serial` reads garbage.
Is that normal behavior?
I'm using Nucleo-L053R8. Have also tried manually override `SystemClock_Config` to use `HSI` but that didn't change the behavior.
**UPD.**
On a long lines it catches first, misses next 4, and the rest is ok at 38400 bauds.
Or catches first, misses next one, and the rest is ok if 9600 bauds.
**UPD2.**
can it be related to this?

> **Low-power sleep mode**
> This mode is achieved by entering Sleep mode with the internal voltage regulator in 
> low-power mode to minimize the regulator’s operating current. In Low-power sleep 
> mode, both the clock frequency and the number of enabled peripherals are limited; a 
> typical example would be to have a timer running at 32 kHz

@fpistm

**Comment by tshcherban** [comment]
Also found that a `STM32LowPower::sleep` does switching the system clock to a MSI (range 1, 131.072 KHz), that explains Serial scramble:
```
#if defined(PWR_LOWPOWERREGULATOR_ON) && \
    (defined(PWR_CSR_REGLPF) || defined(PWR_SR2_REGLPF))
  // When LowPower regulator sleep mode is used, it is necessary to decrease CPU Frequency
  if (regulator == PWR_LOWPOWERREGULATOR_ON) {
    SystemClock_Decrease();
  }
#endif
```
However I couldn't find that within datasheet or a reference manual. RM says that Vdd domain should be not affected
![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/13539236/2efefd18-b647-4a29-b3fe-f5822108c131)

and HSI16 should be disabled only in stop/standby modes
![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/13539236/8e3b4569-bde3-4010-a577-4e204d0bd0f9)

**Reply by fpistm** [reply]
Which board you used ?
Assuming the STM32L051C8T6 (like for the OneWire discussion), errata sheet says:
![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/20641798/acb3659d-9668-4e3d-900a-bdf9e589f604)

And which USART you used or LPUART1 ?

**Reply by tshcherban** [reply]
I'm using STM32L053R8, USART2 and low power sleep mode, neither Stop not Standby. UART is able to wake cpu up successfully, but fails to read the data.
According to the RM if using sleep mode - `LowPower` library should not change the clock to MSI and regulator to lowpower, they should work in lowpower run, sleep and lowpower sleep modes.

**Reply by fpistm** [reply]
Well switch to MSI is to be as generic as possible and not specifically linked to UART.
So, this is normal you lost some byte while the wakeup occurs. Simply adapt to sent dummy frame to wake up the board then sent you data.

**Reply by tshcherban** [reply]
after using `sleep()` even once in the setup - Serial is broken, cant read reliably at all. I've tried different speeds, and it only changes error pattern, but no valid data can be read reliably.

**Reply by fpistm** [reply]
Is it a custom board ?

**Reply by tshcherban** [reply]
nope, that's a Nucleo-L053R8.

**Reply by fpistm** [reply]
OK.
I've tested, some data are effectively lost but this is due to the fact the MCU enter in deepsleep again before receiving all data.
Could you try this sketch updated, Only 1 bytes is lost due to the wakeup time.
```C++
#include "STM32LowPower.h"

// Declare it volatile since it's incremented inside an interrupt
volatile int wakeup_counter = 0;

void setup() {
  Serial.begin(9600);
  // initialize digital pin LED_BUILTIN as an output.
  pinMode(LED_BUILTIN, OUTPUT);
  // Configure low power
  LowPower.begin();
  // Enable UART in Low Power mode wakeup source
  LowPower.enableWakeupFrom(&Serial, SerialWakeup);
  Serial.println("Start deep sleep wakeup from Serial");
}

void loop() {
  digitalWrite(LED_BUILTIN, HIGH);
  delay(500);
  digitalWrite(LED_BUILTIN, LOW);
  delay(500);
  // Triggers an infinite deep sleep
  // (the device will be woken up only by the registered wakeup sources)
  // The power consumption of the chip will drop consistently
  if (!Serial.available()) {
    LowPower.deepSleep();
    Serial.print(wakeup_counter);
    Serial.println(" wake up");
  }
  // Empty Serial Rx
  while (Serial.available()) {
    char c = Serial.read();
    Serial.print(c);
  }
  Serial.println();
  Serial.flush();
}

void SerialWakeup() {
  // This function will be called once on device wakeup
  // You can do some little operations here (like changing variables
  // which will be used in the loop)
  // Remember to avoid calling delay() and long running functions
  // since this functions executes in interrupt context
  wakeup_counter++;
}
```

**Reply by tshcherban** [reply]
will try it a bit later, didn't got a chance to verify it

**Reply by uzi18** [reply]
@tshcherban you can always add one or two dummy bytes as first in packet :)
